lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py

```diff
@@ -82,152 +89,140 @@
+    if len(subnets) > 0 and 'state' in subnets[0] and subnets[0]['state'] == "available":
+        return subnets[0]
     else:
         return False
 
 
-def create_subnet(vpc_conn, vpc_id, cidr, az, check_mode):
+def create_subnet(conn, module, vpc_id, cidr, az, check_mode):
     try:
-        new_subnet = vpc_conn.create_subnet(vpc_id, cidr, az, dry_run=check_mode)
+        new_subnet = get_subnet_info(conn.create_subnet(VpcId=vpc_id, CidrBlock=cidr, AvailabilityZone=az, DryRun=check_mode))
         # Sometimes AWS takes its time to create a subnet and so using
         # new subnets's id to do things like create tags results in
         # exception.  boto doesn't seem to refresh 'state' of the newly
         # created subnet, i.e.: it's always 'pending'.
         subnet = False
         while subnet is False:
-            subnet = subnet_exists(vpc_conn, new_subnet.id)
+            subnet = subnet_exists(conn, new_subnet['id'])
             time.sleep(0.1)
-    except EC2ResponseError as e:
-        if e.error_code == "DryRunOperation":
+    except ClientError as e:
+        if e.response['Error']['Code'] == "DryRunOperation":
             subnet = None
-        elif e.error_code == "InvalidSubnet.Conflict":
-            raise AnsibleVPCSubnetCreationException("%s: the CIDR %s conflicts with another subnet with the VPC ID %s." % (e.error_code, cidr, vpc_id))
         else:
-            raise AnsibleVPCSubnetCreationException(
-                'Unable to create subnet {0}, error: {1}'.format(cidr, e))
+            module.fail_json(msg='Unable to create subnet: {0}'.format(e))
 
     return subnet
 
 
-def get_resource_tags(vpc_conn, resource_id):
-    return dict((t.name, t.value) for t in
-                vpc_conn.get_all_tags(filters={'resource-id': resource_id}))
-
-
-def ensure_tags(vpc_conn, resource_id, tags, add_only, check_mode):
+def ensure_tags(conn, module, subnet, tags, add_only, check_mode):
     try:
-        cur_tags = get_resource_tags(vpc_conn, resource_id)
-        if cur_tags == tags:
-            return {'changed': False, 'tags': cur_tags}
+        cur_tags = subnet['tags']
 
         to_delete = dict((k, cur_tags[k]) for k in cur_tags if k not in tags)
         if to_delete and not add_only:
-            vpc_conn.delete_tags(resource_id, to_delete, dry_run=check_mode)
+            conn.delete_tags(Resources=[subnet['id']], Tags=ansible_dict_to_boto3_tag_list(to_delete), DryRun=check_mode)
 
         to_add = dict((k, tags[k]) for k in tags if k not in cur_tags or cur_tags[k] != tags[k])
         if to_add:
-            vpc_conn.create_tags(resource_id, to_add, dry_run=check_mode)
+            conn.create_tags(Resources=[subnet['id']], Tags=ansible_dict_to_boto3_tag_list(to_add), DryRun=check_mode)
+
+    except ClientError as e:
+        if e.response['Error']['Code'] != "DryRunOperation":
+            module.fail_json(msg='Unable to update tags for {0}: {1}'.format(subnet['id'], e))
```

Comment:
Use the [boto3 exception guidelines](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-2)